### PR TITLE
Configure sidecar proxy mode per endpoint

### PR DESCRIPTION
**Pull request by rlenglet**
Add label `io.cilium.k8s.policy.istiosidecarproxy=true` to every endpoint injected by Istio with a Cilium-compatible sidecar.

Remove unused `LabelsHash` field from `Endpoint`.

Replace any use of the `sidecar-http-proxy` flag with the per-endpoint setting based on the label.
Log a deprecation warning if the `sidecar-http-proxy` flag is set

Remove the `sidecar-http-proxy` setting from Kubernetes example YAML files.
Remove the use of the `sidecar-http-proxy` setting in the Istio GSG.

Signed-off-by: Romain Lenglet <romain@covalent.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4523)

### CODE REVIEW

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-docs-please

**Comment by rlenglet** [comment]
test-docs-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
I'm testing this by running the Istio GSG manually.
Only the 3 pods that have been injected by Istio with the Cilium image get the `k8s:io.cilium.k8s.policy.istiosidecarproxy=true` label:
```
$ kubectl exec cilium-dpfrh -n kube-system -ti -- cilium endpoint list
ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])                                                    IPv6                 IPv4            STATUS   
           ENFORCEMENT        ENFORCEMENT                                                                                                                                      
...
3661       Enabled            Disabled          41695      k8s:app=reviews                                                                f00d::a0f:0:0:e4d    10.15.57.68     not-ready   
                                                           k8s:io.cilium.k8s.policy.istiosidecarproxy=true                                                                                 
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default                                                                                 
                                                           k8s:io.kubernetes.pod.namespace=default                                                                                         
                                                           k8s:version=v1                                                                                                                  
...
45872      Enabled            Disabled          28517      k8s:app=details                                                                f00d::a0f:0:0:b330   10.15.92.101    ready       
                                                           k8s:io.cilium.k8s.policy.istiosidecarproxy=true                                                                                 
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default                                                                                 
                                                           k8s:io.kubernetes.pod.namespace=default                                                                                         
                                                           k8s:version=v1                                                                                                                  
...
61689      Enabled            Disabled          56501      k8s:app=productpage                                                            f00d::a0f:0:0:f0f9   10.15.17.185    not-ready   
                                                           k8s:io.cilium.k8s.policy.istiosidecarproxy=true                                                                                 
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default                                                                                 
                                                           k8s:io.kubernetes.pod.namespace=default                                                                                         
                                                           k8s:version=v1                                                                                                                  
```
Envoy is not started in the Cilium pod:
```
$ kubectl exec cilium-dpfrh -n kube-system -ti -- ps -ef | grep envoy | wc -l
       0
```
```
$ kubectl logs cilium-dpfrh -n kube-system --timestamps | grep 'Envoy: Starting:' | wc -l
       0
```
L7 policy enforcement works as expected in the sidecar:
```
$ for APIPATH in /api/v1/products /api/v1/products/0 /api/v1/products/0/reviews /api/v1/products/0/ratings; do echo ; curl -s -S "${PRODUCTPAGE}${APIPATH}" ; echo ; done

[{"descriptionHtml": "<a href=\"https://en.wikipedia.org/wiki/The_Comedy_of_Errors\">Wikipedia Summary</a>: The Comedy of Errors is one of <b>William Shakespeare's</b> early plays. It is his shortest and one of his most farcical comedies, with a major part of the humour coming from slapstick and mistaken identity, in addition to puns and word play.", "id": 0, "title": "The Comedy of Errors"}]

{"publisher": "PublisherA", "language": "English", "author": "William Shakespeare", "id": 0, "ISBN-10": "1234567890", "ISBN-13": "123-1234567890", "year": 1595, "type": "paperback", "pages": 200}

Access denied


Access denied

```

**Comment by rlenglet** [comment]
Successful CI builds:
https://jenkins.cilium.io/job/Cilium-PR-Doc-Tests/260/
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/4161/
Rebasing...

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-docs-please

File: pkg/workloads/kubernetes.go [hunk 1]
```diff
@@ -69,5 +88,55 @@ func fetchK8sLabels(containerLbls map[string]string) (map[string]string, error)
 		delete(k8sLabels, k8sConst.PolicyLabelServiceAccount)
 	}
 
+	// If the pod has been injected with an Istio sidecar proxy compatible with
+	// Cilium, add a label to notify that.
+	// If the pod already contains that label to explicitly enable or disable
+	// the sidecar proxy mode, keep it as is.
+	if _, ok := k8sLabels[k8sConst.PolicyLabelIstioSidecarProxy]; !ok &&
```
Comment (tgraf) [inline pkg/workloads/kubernetes.go:95]: Should this check for "false" besides non-existing label as well?

File: daemon/main.go [hunk 2]
```diff
@@ -693,6 +693,10 @@ func initEnv(cmd *cobra.Command) {
 	}
 
 	log.Infof("Container runtime options set: %s", workloads.GetRuntimeOptions())
+
+	if viper.GetBool("sidecar-http-proxy") {
```
Comment (tgraf) [inline daemon/main.go:697]: You can use `flags.MarkHidden("sidecar-http-proxy")` to hide it from usage but still support setting the flag.

File: Documentation/gettingstarted/istio.rst [hunk 3]
```diff
@@ -9,114 +9,7 @@ environment running on your machine.
...
-
-::
-
-    $ kubectl get daemonsets -n kube-system
-    NAME      DESIRED   CURRENT   READY     UP-TO-DATE   AVAILABLE   NODE-SELECTOR   AGE
-    cilium    1         1         0         1            0           <none>          6s
-
-Wait until the cilium Deployment shows a ``CURRENT`` count of ``1``
-like above (a ``READY`` value of ``0`` is OK for this tutorial).
+.. include:: cilium_install.rst
```
Comment (tgraf) [inline Documentation/gettingstarted/istio.rst:12]: Awesome!

**Review by tgraf — DISMISSED** [review]
Looks great

File: daemon/main.go (hunk 2 above)
Comment (rlenglet) [reply daemon/main.go:697]: That's always been the case for this flag.

File: pkg/workloads/kubernetes.go (hunk 1 above)
Comment (rlenglet) [reply pkg/workloads/kubernetes.go:95]: Right. Initially I defined this label as having no value. I didn't update this test.

Comment (rlenglet) [reply pkg/workloads/kubernetes.go:95]: Oh, actually this is the behavior I intended. If the user knows better than Cilium can guess, and sets the label to either `"false"` or `"true"`, we don't overwrite that value.

Comment (tgraf) [reply pkg/workloads/kubernetes.go:95]: 👍 makes sense. Just needs to be documented then.

**Review by tgraf — APPROVED** [review]

**Pull request merged**